src/deploy/functions/triggerRegionHelper.ts

```diff
@@ -2,50 +2,72 @@ import * as backend from "./backend";
 import * as storage from "../../gcp/storage";
 import { FirebaseError } from "../../error";
 import { logger } from "../../logger";
+import { regionInLocation } from "../../gcp/location";
 
 const noop = (): Promise<void> => Promise.resolve();
 
-const LOOKUP_BY_EVENT_TYPE: Record<string, (ep: backend.EventTriggered) => Promise<void>> = {
+const ENSURE_BY_EVENT_TYPE: Record<
+  string,
+  (ep: backend.EventTriggered, r: string) => Promise<void>
+> = {
   "google.cloud.pubsub.topic.v1.messagePublished": noop,
-  "google.cloud.storage.object.v1.finalized": lookupBucketRegion,
-  "google.cloud.storage.object.v1.archived": lookupBucketRegion,
-  "google.cloud.storage.object.v1.deleted": lookupBucketRegion,
-  "google.cloud.storage.object.v1.metadataUpdated": lookupBucketRegion,
+  "google.cloud.storage.object.v1.finalized": ensureStorageTriggerRegion,
```

Comment:
This is the second time we have a lookup map from event type -> strategy function (that varies only by service).

I wonder if a cleaner way of understanding the codebase as a whole is to have a lookup function from event type => service delegate.

So something like:

```
// deploy/functions/services/index.ts
interface Service {
   // E.g. "pubsub", "storage"
   readonly name: string;
  
   // Rough swag here
   missingProjectBindings(projectId: string, existingPolicy: iam.Policy): Promise<Array<iam.Binding>>
   ensureTriggerRegion(): Promise<void>;
}

// Up to you to decide whether to return undefined or a noop-ing service for v1/pub/sub etc
// Though I personally might choose a noop impl to avoid null checks everywhere.
function serviceForEndpoint(endpoint: backend.Endpoint): Service

// deploy/functions/service/storage.ts
class StorageService implements Service {
  ...
};
```

Take this as a "wondering aloud", not a strong nudge. I'm curious for your thoughts on future comprehensibility.

Reply:
That makes a lot of sense and will probably help us out in the future. In the latest commit I've implemented this pattern.